sil.abnf

```diff
@@ -0,0 +1,71 @@
+; Formal grammar definition for SIL (SILE Input Language) files
+;
+; Based on RFC 5234 (Augmented BNF for Syntax Specifications: ABNF)
+; Uses RFC 7405 (Case-Sensitive String Support in ABNF)
+
+; A master document can only have one top level content item, but we allow
+; loading of fragments as well which can have any number of top level content
+; items, hence valid grammar can be any number of content items.
+document = *content
+
+; Top level content can be any sequence of these things
+content =  environment
+content =/ comment
+content =/ text
+content =/ braced_content
+content =/ command
+
+; Environments come in two flavors, raw passthrough and regular. The difference
+; is what is allowed to terminate them and what escapes are needed for the
+; content in the middle.
+environment =  %s"\begin" [ options ] "{" command-id "}" content %s"\end{" command-id "}"
+environment =/ %s"\begin" [ options ] "{" passthrough-command-id "}" passthrough-content %s"\end{" passthrough-command-id "}"
+
+; Nothing to see here.
+comment = "%" utf8-octets CRLF
+
+; Input strings that are not wrapped in some SILE command.
+text = *text-char
+
+; Input content wrapped in braces can be attatched to a command or just used to
+; create a settings wrapper.
+braced_content = "{" content "}"
+
+; As with environments, can be raw passthrough or regular content depending on the command. With raw variants we require contents.
+command =  "\" command-id [ options ] [ braced_content ]
+command =/ "\" passthrough-command-id [ options ] braced-passthrough-content
+
+; Building blocks
+options = "[" parameter *( "," parameter ) "]"
```

Comment:
Current syntax (pairsep in parserbits.lua) also allows the semi-colon as separator

Reply:
Yahoo, who's idea was it to have alternate syntax there? (I'll probably `git blame` and find out it was me). How bad would it be do you think to walk that back and only have one syntax? We just make a long tail of work for future tooling if the parser is full of alternative ways to do things.

Reply:
The SILE manual actually mentions it: "... multiple parameters are separated by commas or semicolons"
But on my own, I would never have the idea of using the semicolumn, and I would agree to only have one syntax eventually. Unless there was a good reason to have it too, we could deprecate it, as far as II am concerned, or keep it undocumented.

Reply:
A little more thought early on and we might have had a way to pass a table or list as an option without quoting (e.g. if commas were used for one and semicolons for another, `\foo[string=foo; table=val1,val2,val3]`) but I suspect that ship has sailed. Unless we reassign the semicolon to be the *lesser* operator which would just rub me and probably most English speakers the wrong way, we're stuck with the quoting method `\foo[string=foo,list="val1,val2,val3"]`.

Having more that one possible separator just makes more work for users needing to know about both and escape them, more work for parsers, editor syntax grammars, etc.

I vote we un-document this pronto as deprecated and remove it in a future release. I might be able to work out a way to throw a deprecation warning from the LPEG grammar, or maybe not.

Reply:
All these months later and I am still of the same mind. We should not have pointless alternate syntax variants: they just make more work all the way around. Hopefully there isn't too much of this in the wild. I wish we'd caught it a lot earlier and could have used the semicolon variant and been able to use commas for multiple values, but I'm pretty sure any attempt to backtrack that is going to make downstream users very unhappy. It just isn't a nice easy thing to shim or reformat.

I un-documented the alternative as a starting point. We'll leave it in the actual parser for a while but look into a way to throw a deprecation warning on it.